ml/src/main/scala/ai/h2o/sparkling/ml/algos/H2OAlgorithm.scala

```diff
@@ -16,98 +16,95 @@
+import ai.h2o.sparkling.model.H2OModel
+import com.google.gson.{Gson, JsonElement}
 import hex.Model
 import org.apache.spark.annotation.DeveloperApi
-import org.apache.spark.h2o._
 import org.apache.spark.ml.Estimator
 import org.apache.spark.ml.param.ParamMap
 import org.apache.spark.ml.util._
 import org.apache.spark.sql.Dataset
 import org.apache.spark.sql.types.StructType
-import water.exceptions.H2OModelBuilderIllegalArgumentException
-import water.support.ModelSerializationSupport
-import water.{DKV, H2O, Key}
 
-import scala.reflect.{ClassTag, classTag}
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
 
 /**
  * Base class for H2O algorithm wrapper as a Spark transformer.
  */
-abstract class H2OAlgorithm[B <: H2OBaseModelBuilder : ClassTag, M <: H2OBaseModel, P <: Model.Parameters : ClassTag]
-  extends Estimator[H2OMOJOModel] with H2OAlgoCommonUtils with DefaultParamsWritable with H2OAlgoCommonParams[P] {
+abstract class H2OAlgorithm[P <: Model.Parameters : ClassTag]
+  extends Estimator[H2OMOJOModel]
+    with H2OAlgoCommonUtils
+    with DefaultParamsWritable
+    with H2OAlgoCommonParams[P]
+    with RestCommunication {
 
-  protected def preProcessBeforeFit(trainFrameKey: String): Unit = {}
+  protected def prepareH2OTrainFrameForFitting(frame: H2OFrame): Unit = {}
 
   override def fit(dataset: Dataset[_]): H2OMOJOModel = {
-    // Update H2O params based on provided configuration
-    updateH2OParams()
-
-    val (trainKey, validKey, internalFeatureCols) = prepareDatasetForFitting(dataset)
-    parameters._train = DKV.getGet[Frame](trainKey)._key
-    parameters._valid = validKey.map(DKV.getGet[Frame](_)._key).orNull
-
-    val trainFrame = parameters._train.get()
-    preProcessBeforeFit(trainFrame._key.toString)
-    if (!RestApiUtils.isRestAPIBased()) {
-      water.DKV.put(trainFrame)
+    val (train, valid, internalFeatureCols) = prepareDatasetForFitting(dataset)
+    prepareH2OTrainFrameForFitting(train)
+    val params = getH2OAlgorithmParams() ++
+      Map(
+        "training_frame" -> train.frameId,
+        "model_id" -> convertModelIdToKey()
+      ) ++
+      valid.map { fr => Map("validation_frame" -> fr.frameId) }.getOrElse(Map())
+    val content = try {
+      RestApiUtils.updateAsJson(s"/3/ModelBuilders/${parameters.algoName().toLowerCase}", params)
+    } catch {
+      case e: RestApiCommunicationException if e.getMessage.contains("There are no usable columns to generate model") =>
+        throw new IllegalArgumentException(s"H2O could not use any of the specified feature" +
+          s" columns: '${getFeaturesCols().mkString(", ")}'. H2O ignores constant columns, are all the columns constants?")
     }
-
-    // Train
-    val binaryModel: H2OBaseModel = trainModel(parameters)
-    val mojoData = ModelSerializationSupport.getMojoData(binaryModel)
+    val gson = new Gson()
+    val job = gson.fromJson(content, classOf[JsonElement]).getAsJsonObject.get("job").getAsJsonObject
```

Comment:
Can't we reuse some H2O V3 objects for reserialization?

Reply:
I will try one more time. I had problems with de-serialization of some fields, but maybe I can be more specific about exclusion rules